File: src/main/java/org/roborace/lapscounter/domain/Robot.java
package org.roborace.lapscounter.domain;

import lombok.*;

import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@ToString
@EqualsAndHashCode(exclude = "lapTimes")
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Robot {

    private int serial;

    private String name;

    private int num;
    private int place;
    private int laps;
    private long time;

    private long currentLapStartTime;
    private long lastLapTime;
    private long bestLapTime;

    private final List<Long> lapTimes = new ArrayList<>();

    public void incLaps(long raceTime) {
        laps++;
        if (laps > 0) {
            lapTimes.add(raceTime);
            time = raceTime;
        }
        lastLapTime = raceTime - currentLapStartTime;
        currentLapStartTime = raceTime;
        if (lastLapTime < bestLapTime || bestLapTime == 0) {
            bestLapTime = lastLapTime;
        }
    }

    public void decLaps() {
        laps--;
    }

    public void reset() {
        laps = 0;
        time = currentLapStartTime = lastLapTime = bestLapTime = 0;
        lapTimes.clear();
    }

    public long extractLastLapTime() {
        if (lapTimes.isEmpty()) return 0L;
        lapTimes.remove(lapTimes.size() - 1);
        if (lapTimes.isEmpty()) return 0L;
        return lapTimes.get(lapTimes.size() - 1);
    }
}

File: src/main/kotlin/org/roborace/lapscounter/domain/Robot.kt
package org.roborace.lapscounter.domain

data class Robot(
    val serial: Int,
    var name: String? = null,
    val num: Int = 0,
    var place: Int = 0,
    var laps: Int = 0,
    var time: Long = 0,
) {
    private val lapTimes: MutableList<Long> = mutableListOf()

    var currentLapStartTime: Long = 0
    var lastLapTime: Long = 0

    var bestLapTime: Long = 0

    var pitStopFinishTime: Long? = null

    fun incLaps(raceTime: Long) {
        laps++
        if (laps > 0) {
            lapTimes.add(raceTime)
            time = raceTime
        }
        lastLapTime = raceTime - currentLapStartTime
        currentLapStartTime = raceTime
        if (bestLapTime == 0L || lastLapTime < bestLapTime) {
            bestLapTime = lastLapTime
        }
    }

    fun decLap() {
        laps--
        lapTimes.removeLastOrNull()
        time = lapTimes.lastOrNull() ?: 0L
    }

    fun reset() {
        laps = 0
        time = 0
        lapTimes.clear()
        bestLapTime = 0
        lastLapTime = 0
        currentLapStartTime = 0
        pitStopFinishTime = null
    }
}
